[2단계 - 로또 구현] 검프(김태정) 미션 제출합니다.

Livenow14 opened the pull request:
안녕하세요 코니! 좋은 아침입니다 😊

보내주신 링크와 피드백을 여러번 읽었습니다!! 

읽으면서 그동안 가졌던 문제를 생각해봣는데요,

우선, 어떠한 기능을 쓸줄만 알지 그 지식의 동작방식 혹은 개념을 정확히 이해하지 않고 대충 넘어가려햇습니다. 

또한 과하게 확장성을 고려하며 코드를 작성했던거 같습니다.

> 그래서 이번에는, 현재에만 집중해보기로 했습니다.

클래스들이 최대한 할 수 있는 일이 무었일까, 가지고 있는 클래스들로만 역할을 수행할 수 있지 않을까 생각해봤습니다. (책임을 분리하는 것도 좋지만, 너무 분리하는 것도 능사는 아닌거 같거든요 😭)

인스턴스의 선언이유, 메소드의 선언 이유에 조금 더 집중하였습니다.

## 리팩토링 목록
### step1 2차 피드백
- [x] GameManageApplication 메서드 순서 변경 - d534d61
- [x] Result 인스턴스 변수 제거 - 18a1327
- [x] InputUtils 책임 재정의 - 562e9c7
- [x] RandomLottoBall 생성 책임 변경하기 - ddb0c9b
- [x] LottoGameMachine 네이밍 변경
- [x] 출력 메시지 동일하게 수정 - cc77469
- [x] HashMap 메소드 사용해서 코드 줄이기 - e1767fc
- [x] 정적 메소드 팩토리 제거 - 9416592
- [x] 네이밍 변경,  하드코딩 한 값 변경 - 46992ad
- [x] LottoRank 테스트 코드 추가 - 46992ad
- [x] Stream.forEach()를 Collections.forEach()로 변환

이번에도 코드를 작성하다 몇가지 의문점이 생겼어요..ㅎㅎ 

## 의문점.

### 첫번째, 수동 로또 넘버는 input을 받아야 테스트 할 수 있는데, 그러다보니 실제 입력을 받야할 곳에 inputView을 넘겨줘야한다. 이렇게 입력을 위한 객체를 메소드를 통해 넘겨주는 것이 맞는지

```java
public class LottoMachine {
...
public List<LottoTicket> makeManualTickets(final InputView inputView, final int ticketCount) {
        if (ticketCount == 0) {
            return Collections.emptyList();
        }
        OutputUtil.printMessage("수동으로 구매할 번호를 입력해 주세요.");
        return IntStream.range(0, ticketCount)
                .mapToObj(count -> makeManualTicket(inputView))
                .collect(Collectors.toList());
    }

    private LottoTicket makeManualTicket(final InputView inputView) {
        List<Integer> manualTicketNumbers = inputView.inputManualTicketNumber();
        List<LottoBall> lottoBalls = manualTicketNumbers.stream()
                .map(LottoBall::new)
                .collect(Collectors.toList());
        return new LottoTicket(new LottoBalls(lottoBalls));
    }
...
}
```

현재 LottoMachine에서 수동 티켓을 발급하는 책임을 지니고 있는데요, 현재와 같이 입력의 부분이 들어가니 InputView를 전달인자로 전달하는 것을 볼 수 있습니다. 

> 그래서 질문은,

input output은 최대한 controller에서 진행해야한다고 생각하는데, 그렇다면 수동 로또 입력의 경우도 필요한 입력을 controller에서 다 받고, 입력값 자체를 LottoMachine에게 보내는게 맞는것일까요?

### ~~두번째, 싱글톤의 사용시기~~

~~`1~45개의 로또가 계속 생성됐다 버려지니, 캐싱을 해놓는게 좋지 않을까?` 라는 피드백을 생각하며, LottoBallFactory라는 싱글톤 객체를 만들었는데요. (만들어진 로또 번호를 반환하는 책임)~~

~~"이러한 작업이 현재 애플리케이션에서 과할까?" 싶기도 하고, "만들어진 로또번호를 저장할 곳이 마땅한가"에 대해 명확한 답을 못 내리겠더라고요..😂~~ 

> ~~그래서 질문은,~~

~~제가 생각하는 싱글톤의 사용이유는 해당 클래스의 인스턴스가 1개만 만들어지도록 보장하며, 사용하는 여러 곳에서 1개의 인스턴스를 공유하며 재사용 할 때 필요하다 생각합니다. (공유하기 때문에 상태를 가지면 안됨)~~

~~다만, 이런 상황이 현재와 같은상황인지 아직 잘 모르겠어요..~~

~~코니의 생각은 어떤가요?~~

## 결론적으로..

코니의 피드백을 통해, `의미없이(나의 명확한 이유없이) 작성했던 코드`들이 다른 사람의 입장에선 모두 의도된 것으로 보이는구나라는 생각을 많이 했던 1주일 이었습니다. →( 이부분에서 정말 코드 한줄 한줄 이유가 있어야하는 것을 알게되었어요 )

즉, 코드 작성시의 마음가짐과 자세를 돌아보게 되었습니다( 아직 부족하지만, 의식적으로 가지길 노력할게요 😊 )

코니의 피드백 덕분에 정말 많은 것을 배울 수 있었어요. 감사합니다 🙇🏻‍♂️

이번 피드백도 잘 부탁드립니다 :)

Livenow14 committed:
refactor: GameManageApplication 메서드 순서 변경

가독성을 위해 메스드 명명 순서를 변경함

Livenow14 committed:
refactor: InputUtils 책임 재정의

Livenow14 committed:
refactor: LottoGameMachine 네이밍 변경

- [x] RandomLottoBall 생성 책임 변경 -> LottoMachine으로 이동

Livenow14 committed:
refactor: 출력 메시지 동일하게 수정

전체 메시지리를 상수화 하여 클래스 접근시 바로 확인할 수 잇게 함.

Livenow14 committed:
refactor: HashMap 메소드를 통한 변경

코드 중복을 부분 변경하였음

Livenow14 committed:
refactor:  LottoTickets 리팩토링 및 테스트 진행

Livenow14 committed:
refactor: 정적 메소드 팩토리 제거

생성자를 통해 그 의도를 잘  드러나게 변경하였음

Livenow14 committed:
refactor: LottoRank 테스트 코드 추가

- [x] 네이밍 변경,  하드코딩 한 값 변경
- [x] LottoRank 테스트 코드 추가

Livenow14 committed:
refactor:Stream.forEach()를 Collections.forEach()로 변환

Livenow14 committed:
refactor: Collectors.joining 사용

-lottoRankTest 테스트 케이스 추가
- LottoBallFactory 싱글톤으로 변경

Livenow14 committed:
refactor: Result 인스턴스 변수 제거

생성시에만 사용되는 변수 제거 및 반환값이 필요없는 매소드 변경

Livenow14 committed:
refactor: 메소드 네이밍 변경

Livenow14 committed:
refactor: InputUtil 책임 분리

LottoGameScreen이 가지고있던 입력할때 필요한 메세지 InputUtills로 옮김

Livenow14 committed:
docs: 새로운 요구사항 정리

Livenow14 committed:
refactor: BettingMoney 정적 팩토리 메소드 제거

인자값과 생성자의 인자값이 같다고 판단되어 제거함.

Livenow14 committed:
feat: 구매 갯수만큼의 랜덤 로또 생성 로직 추가

LottoMachine -> RandomLottoMachine
buyTickets -> 제거 후 makeTickets 추가

Livenow14 committed:
feat 수동으로 구매할 보장된(양수이며, 구입가능 수보다 작거나 같은 자연수) 로또 수를 입력받는다.

테스트 코드작성 완료

Livenow14 committed:
feat: 수동으로 구매 할 보장된 로또 번호를 입력받는다.

InputUtil -> InputView로 수정
RandomLottoMachine -> LottoMachine 수정
수동 구매  로직 테스트

Livenow14 committed:
feat: 전체 티켓 중 수동으로 구매하지 않은 티켓 랜덤으로 생성한다.

수동 구매한 로또 콘솔에 출력된다.
랜덤 구매한 로또 콘솔에 출력된다.

Livenow14 committed:
fix: 수동구매 티켓이 0개일시 발생하는 에러 수정

0임을 체크하는 로직 추가

Livenow14 committed:
fix: 22억 이상의 상금일 시 나타내는 에러 수정

상금이 int의 최대 값을 넘어가면 생기는 에러 수정
Long으로 변경하엿음.

Livenow14 committed:
refactor: TickeCount 예외 발생 조건 변경

0 이하 -> 0 미만으로 예외 발생을 변경시킴

Livenow14 committed:
refactor: final 키워드 추가

매개변수에 final이 빠진부부 추가하였음

Livenow14 committed:
refactor: 입력 책임 controller로 변경

Livenow14 commented:
## 검프의 학습 로그 
## **자바 상속의 특징** - 1

### **상속(Inheritance)**

부모가 소유하고 있는 재산의 일부를 자식이 물려받는 것을 생각해보면, 부모가 가진 것을 자식이 `마음대로 쓰거나` `그대로 쓰거나 선택`할 수 잇습니다. 

즉, `자식 클래스가 부모 클래스의 변수와 메소드를 물려 받아 쓰는 것`입니다.

### **상속의 장점**

코드의 재사용을 통해 코드의 간결성을 확보할 수 있습니다.

### **상속의 단점**

부모 클래스 기능에 버그가 생기거나 기능의 추가/변경 등으로 변화가 생겼을 때, `자식 클래스가 정상적으로 작동할 수 있을지에 대한 예측이 힘듭`니다.

자식 클래스가 동작하기 위해선 부모 클래스에 의존할 수 밖에 없습니다.

 즉, 부모 클래스와의 결합도가 높아집니다.(**캡슐화 원칙**에 위반합니다). 

### **상속의 특징**

자식 클래스는 단 하나의 부모 클래스로부터 상속 받을 수 있습니다. (다중 상속 금지)

→ `인터페이스`는 다중 상속 가능합니다. (헷갈려서 한번더 정의하고 가겠습니다.)

상속받는 자식 클래스가 다른 클래스의 부모 클래스가 될 수 있습니다.

런타임 때, 자식클래스의 오버라이딩 한 변수나 메소드가 우선적으로 사용됩니다. 
`접근제어자`를 통해 상속의 범위를 조정할 수 있습니다.

### **사용법**

```java
# 나쁜 예
class 자식클래스 extends 부모클래스, 다른부모클래스 {
}

# 좋은 예
class 자식클래스 extends 부모클래스 {
}
```

Livenow14 commented:
## 리퍼털이란? - 0.5
## 리터럴

**프로그램에서 직접 표현한 값,**

쉽게 얘기해서 소스 코드(변수에 넣는 데이터)의 고정된 값을 의미하는 용어입니다.

아래의 예제에서

```java
int a = 1;
```

int 앞에 a는 `변수`이고, 여기서의 1은 `리터럴` 을 의미합니다.

즉, 1과 같이 변하지 않는 데이터를 리터럴이라고 합니다.

### 리터럴의 종류는 정수, 실수, 논리, 문자, 문자열 리터럴 등이 있습니다.

### 정수형 리터럴

정수형 리터럴은 일반적인 숫자 즉 정수 데이터를 의미합니다

- 10진수 리터럴

    ```java
    int a = 15;
    ```

    - 10진수 값으로 10을 출력
- 8진수 리터럴

    ```java
    int b = 015;
    ```

    - 10진수 값으로 13 출력
- 16진수 리터럴

    ```java
    int c = ox15;
    ```

    - 10진수 값으로 21 출력
- 2진수 리터럴

    ```java
    int d = 0b0101;
    ```

    - 10진수 값으로 5 출력

> 모든 정수형 데이터가 기본적으로 int형이기 때문에 long 데이터 자료형에 정확한 long 리터럴을 지정하기 위해서는 숫자 뒤에 알파벳 l(또는 L)를 추가 해줘야 합니다.

### 실수(부동 소수점) 리터럴

소수점을 가진 실수형 데이터를 의미합니다

```java
double a = 1.234; // 일반적인 실수 표현 방식
double b = 6.02E23; // 지수를 이용한 표현 방식 
float c = 1.234F; // float형 표현 방식
```

> 모든 실수형 데이터는 기본적으로 double이기 때문에 float 데이터 자료형에 정확한 float 리터럴을 지정하기 위해서는 숫자뒤에 알파벳 f(또는 F)를 추가 해줘야 합니다.

### 논리 리터럴

참(true)과 거짓(false)을 표현할때 사용하는 논리데이터를 의미합니다.

```java
boolean a = true;
boolean b = 10 > 0;
```

> 1, 0 으로 참 거짓을 표현할 수 없습니다.

### 문자 리터럴

자바의 모든 문자들은 Unicode를 사용합니다. 정수로 변환될 수 있으며, 더하고 빼는 거와 같은 연산도 가능합니다. 정수 형태가 아닌 **문자 리터럴을 표시하고자 할 때는 단일 인용 부호(' ')를 사용**합니다. 유니코드나 직접 입력이 불가능한 문자들에 대해서는 역슬래쉬( \ )를 이용하여 표시할 수 있습니다.

```java
char a = 'L';
char b = "삶";
char c = \uae00;(유니코드값)
```

![https://user-images.githubusercontent.com/48986787/104325771-56e98a80-552c-11eb-859e-c3d274ef4093.png](https://user-images.githubusercontent.com/48986787/104325771-56e98a80-552c-11eb-859e-c3d274ef4093.png)

### 문자열 리터럴

**문자열 리터럴을 표시하고자 할 때는 이중 인용 부호("")를 사용**합니다.

```java
String str = "안녕?";
```

> 문자열 리터럴은 기본(프리미티브) 타입이 아닙니다.

### null 리터럴

레퍼런스 타입에 대입해서 사용합니다. (기본 타입에는 사용불가)

```java
int a = null; // 에러가 난다. 
String str = null;
str = "JAVA";
```

Livenow14 committed:
refactor: LottoRank 클래스 상수 네이밍 번경 및 인스턴스 네이밍 변경

hasBonus -> isSecond

Livenow14 committed:
refactor: LottoBallFactory 삭제 및 책임 이동

LottoBall에서 static 한 Map을 가지게 변경하였음.
그로 인한 변경들을 수정함.

choihz commented on src/main/java/Application.java:
```diff
@@ -1,13 +1,15 @@
-        GameManageApplication gameManageApplication = new GameManageApplication(new MainScreen(), new LottoGameScreen(), lottoService);
+        LottoMachine lottoMachine = new LottoMachine();
+        LottoService lottoService = new LottoService(lottoMachine);
+        GameManageApplication gameManageApplication = new GameManageApplication(new LottoGameScreen(), lottoService, new InputView(new Scanner(System.in)));
```
여기서부터 이렇게 다 생성해 줄 필요가 있나요? 필요한 곳에서 생성해도 되지 않나요?

Livenow14 replied:
애플리케이션에서 사용영역과 구성영역을 봤을때, 
Application은 사용영역 보단 애플리케이션의 원활한 사용을 위한 구성영역이라 생각했습니다.
현재 controller의 역할을 하는 GameManageApplication이 가지는 인스턴스들은 메소드에서 실 사용시 구성의 역할이 없게, 즉 사용영역에서는 어떤 객체가 들어온지 몰라도 사용시 문제가 없게 만들어주기 위해(이미 구성되어있는 객체를 사용하기 때문에) 현재와 같이 선언했습니다. 

혹시 메서드에서 실 사용시 인스턴스화 하는 것이 Controller의 역할이 더 잘 드러날까요? 
(LottoService는 현재 run()메소드에서만 사용되기 때문에 인스턴스 변수로 두는게 어색해 보였을까요?)

choihz replied:
"사용영역에서는 어떤 객체가 들어온지 몰라도 사용시 문제가 없게 만들어주기 위해"라고 하셨는데 어차피 객체들의 타입은 정해져 있는 상황이고(인터페이스가 아니죠), `GameManageApplication`을 사용하는 클라이언트인 `Application` 입장에서는 필요한 객체들을 다 알고 만들어 넣어줘야 하는 상황인데 이 로또 어플리케이션의 맥락에서는 이러한 일이 불필요해 보여 남긴 코멘트였습니다

choihz commented on src/main/java/domain/ball/LottoBall.java:
```diff
@@ -1,27 +1,38 @@
-    private boolean isBetweenNumber(final int number) {
-        return number >= MIN_LOTTO_VALUE && number <= MAX_LOTTO_VALUE;
+    public static List<LottoBall> getLottoBalls() {
+       return lottoBalls.values().stream().collect(Collectors.toList());
```
```suggestion
       return new ArrayList<>(lottoBalls.values());
```

Livenow14 replied:
`Collection으로 인스턴스화 된 객체를 List로 다운캐스팅 하는 방법을 제대로 몰랐어요`.
List구현체로도 충분히 할 수 있는 일을 stream으로 한번더 하게되어, 메모리에 낭비되는 Stream객체가 생겼겠군요!!  
변경 후 더 깔끔한 코드가 되었습니다 감사합니다 😁 
ed95636

choihz commented on src/main/java/domain/lotto/TicketCount.java:
```diff
@@ -1,27 +1,50 @@
         return ticketCount;
     }
 
     private void validateLottoCount(final int lottoCount) {
```
생성자에서 매개변수로 넘겨주는 건 `ticketCount`, 여기선 `lottoCount`라서 헷갈리네요 🙄

Livenow14 replied:
ㅇ으아.. 네이밍의 통일성을 지키지 못했네요...ㅠㅠ
확실히 모르는 사람이 코드를 봤을때 헷갈렸을거 같아요..ㅎㅎ

좋은 지적 감사합니다.😊

choihz commented on src/main/java/GameManageApplication.java:
```diff
@@ -4,53 +4,49 @@
+        return new BettingMoney(input);
     }
 
     private TicketCount getTicketCount(final BettingMoney bettingMoney) {
```
이 작업을 여기서 해줄 필요가 있을까요? 지금 로직에선 `bettingMoney.getTicketCount()`가 바로 `new TicketCount(ticketCount)`를 리턴할 수도 있을 것 같아요.

Livenow14 replied:
getTicketCount가 바로 TicketCount 객체를 반환해 주는 방법이 있었네요!! 
BettingMoney가 가지고 있던 `LottoTicket.TICKET_PRICE를 없애고`, `getTicketCount에게 값을 넘겨주는 방식`으로 로직을 변경햇습니다. 
결론적으로 BettingMoney가 알고있던 TICKET_PRICE를 없앨 수 있었어요..! 감사합니다.


3f517a8

choihz commented on src/main/java/service/LottoService.java:
```diff
@@ -1,21 +1,35 @@
 
-    public LottoTickets getLottoTickets(BettingMoney bettingMoney) {
-        List<LottoTicket> lottoTickets = lottoGameMachine.buyTickets(bettingMoney);
+    public LottoTickets getLottoTickets(final InputView inputView, final TicketCount ticketCount, int manualTicketCount) {
```
1. 1단계에서도 말씀드렸던 것처럼, 이 클래스는 책임이 애매해 보여요.
> LottoMachine이 Tickets까지 알고 있어야 할까?라는 생각으로 나온 결과에요.

→ 현재 `LottoMachine`은 사실상 다 알고 있는 것처럼 보이구요,

> 사실상 현재 도메인 로직의 제일 상단에 있다고 볼수있는 LottoTickets를 Machine에서 관리하자니 책임이 커지는 거 같고

→ 어떤 이유로 책임이 '크다'고 느끼셨을까요?

> controller에서 관리하자니 도메인 자체를 만지는게 어색해보였어요

→ 책임을 더 잘 나눠 보면 어떨까요?

2. `InputView`가 `GameManageApplication`, `LottoService`, `LottoMachine`에게까지 전달되고 있네요. `LottoMachine`은 `InputView`와 `OutputUtil`을 직접 사용하고 있구요.
만약 기존의 콘솔 프로그램을 그대로 유지하면서 웹뷰도 지원해야 한다는 요구사항이 추가되었다고 가정해 볼까요? (물론 제가 `확장 가능성과 수정 가능성을 염두에 두는 것은 분명 중요하지만, 미리부터 일어나지 않을 일을 지나치게 고려할 필요는 없더라구요~` 이런 말을 하긴 했지만요 😅) 웹뷰 지원을 위한 수정 범위는 과연 어디까지가 될까요?
질문하신 것을 보면 검프도 어색함(?)을 느끼신 것 같은데, 그럼에도 이렇게 구현하신 이유가 무엇일까요?

Livenow14 replied:
1. 

```java
1단계에서도 말씀드렸던 것처럼, 이 클래스는 책임이 애매해 보여요.

> LottoMachine이 Tickets까지 알고 있어야 할까?라는 생각으로 나온 결과에요.

→ 현재 `LottoMachine`은 사실상 다 알고 있는 것처럼 보이구요,
```

**re: 그렇다면 LottoMachine에서 LottoTickets또한 조합하여 만들어 주는게 좋다고 생각하면 될까요?**

```java
> 사실상 현재 도메인 로직의 제일 상단에 있다고 볼수있는 LottoTickets를 Machine에서 관리하자니 책임이 커지는 거 같고

→ 어떤 이유로 책임이 '크다'고 느끼셨을까요?
```

**re: LottoTickets를 만드는 거 까지 해야할까? 라는 생각이 책임이 크다고 생각한 거같아요.** Tickets를 만들기 위해 필요한 객체들은 LottoMachine에서 만들어주고, LottoMachine에서 반환해주는 객체들을 이용하여 LottoService에서 필요한 값들을 생성하여 Controller에 주는게 더 자연스러워 보였어요.

```java
> controller에서 관리하자니 도메인 자체를 만지는게 어색해보였어요

→ 책임을 더 잘 나눠 보면 어떨까요?
```
**re: 책임이라는 것이 메소드 관점에서 말씀하시는 건지, 협력의 관점에서 객체가 가져야할 책임을 말씀하시는건지 알 수 있을까요?**

### 메소드의 책임을 나눈다는 관점이라면

controller에서 도메인을 만지는 책임을 controller안의 메소드가 가지게 하면 될거 같아요.

### 협력에서 객체들의 책임을 나눈다는 관점이라면,

LottoService라는 객체말고 이외의 객체가 더 필요하다 생각하면 될까요? 아니면, 현재 협력에 참가하고 있는 객체들안에서 책임을 더 명확하게 할 필요가 있다고 생각하면 될까요?? 

질문들을 정리해보니 LottoMachine에서 현재 LottoService가 가지고 있는 책임을 가져도 괜찮을 거 같다는 생각이 드네요. 

> 변경 전, 궁금증

코니가 질문해주신 것은 어쩌면 LottoMachine이 충분히 가져도 될 책임을 LottoService에 할당함으로써 오히려 협력이 더 흐려졌음을 지적해 주신걸까요..??

Livenow14 replied:
```
InputView가 GameManageApplication, LottoService, LottoMachine에게까지 전달되고 있네요. LottoMachine은 InputView와 OutputUtil을 직접 사용하고 있구요.
만약 기존의 콘솔 프로그램을 그대로 유지하면서 웹뷰도 지원해야 한다는 요구사항이 추가되었다고 가정해 볼까요? (물론 제가 확장 가능성과 수정 가능성을 염두에 두는 것은 분명 중요하지만, 미리부터 일어나지 않을 일을 지나치게 고려할 필요는 없더라구요~ 이런 말을 하긴 했지만요 😅) 웹뷰 지원을 위한 수정 범위는 과연 어디까지가 될까요?
질문하신 것을 보면 검프도 어색함(?)을 느끼신 것 같은데, 그럼에도 이렇게 구현하신 이유가 무엇일까요?

```

구현사항 중 

```java
수동으로 구매할 번호를 입력해 주세요.
8, 21, 23, 41, 42, 43
3, 5, 11, 16, 32, 38
7, 11, 16, 35, 36, 44
```

"로또 번호 부분을 한줄 씩 받을까, controller에서 일단 값들을 다 받고 넘겨줄까"라는 고민의 결과로 나온 구현이에요. inputview를 controller에서 service로 넘겨주면 inputview를 편하게 쓸 수 있지 않을까 했어요.

`다만 코니가 말씀하신 것처럼 어색함이 너무 크게 느껴지네요.. ㅎㅎ`

그래서, 이번 리팩토링에선 controller에서 값들을 우선 다 받고 그 값들을 책임질 객체에게 보내주는 형식으로 바꾸었습니다.

cd2ed3e

choihz replied:
> 별도로 분리될 만큼의 책임을 가지고 있는지 고민해 볼까요? 보통은 작게 나누는 것이 좋긴 하지만, 때로는 연관된 책임이 나뉘어 있으면 오히려 탐색 비용이 올라가기도 하고, 응집도가 떨어져 한 가지의 수정을 여러 곳에서 해야 되기도 해요.

기존의 `LottoService`가 하는 일이 `List<LottoTicket> manualLottoTickets`와 `List<LottoTicket> randomLottoTickets`를 합쳐 `LottoTickets`를 만드는 일뿐이었는데, 이는 `LottoMachine`과 굉장히 강하게 결합되어 있고 또 특별히 분리될 만큼의 로직도 아니라고 생각해서 드린 말씀이었어요!

Livenow14 replied:
코니의 말씀대로 LottoService를 제거 후 LottoMachine이 하는 일이 더 명확해졌습니다 ㅎㅎ 
> 애플리케이션에 Service는 꼭 있어야해!
이런 고정관념에 박혀있었던거 같아요..ㅎㅎ 
코니 피드백을 통해 그런 고정관념을 깨 버릴 수 있었습니다. 감사합니다 😊

choihz commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,63 @@
+                .map(String::trim)
+                .map(Integer::parseInt)
+                .collect(Collectors.toList());
+    }
```
두 메서드가 아주 유사한 일을 하고 있는 것처럼 보이네요! 하나는 `Set`, 하나는 `List`인 이유가 있나요?

Livenow14 replied:
Set 인터페이스는 중복을 허용하지 않습니다. 따로 중복된 값을 찾는 로직을 작성하는 것 보다, Set 구현체를 활용하여 중복값이 들어올 시 사이즈를 비교하여 유효성 검사를 하려 사용했습니다. (winningLotto 입력 시 유효성 검사를 편하게 하기 위해)

다만 전체적으로 적용을 할까하다, 구현 요구사항에 Collection.shuffle(List<>) 사용(이는 Set을 임시로 List로 변환하여 사용할 수 있음)을 해야하고, 만들어진 로또번호를 정렬을 해서 출력해야 했기 때문에 (이또힌 List로 변환해야함) 몇몇부분은 List, 몇몇부분은 Set으로 작성했습니다. 

```java
//InputView.class
public List<List<Integer>> inputManualTicketNumber(int manualTicketCount) {
        OutputUtil.printMessage(MANUAL_TICKET_BALLS);
        return IntStream.range(0, manualTicketCount)
                .mapToObj(count -> scanner.nextLine())
                .map(this::splitManualTicketBalls)
                .collect(Collectors.toList());
    }
```

이부분이 List인 이유는 반환값을 사용하는 

```java
// LottoMachine.class
private LottoTicket makeManualTicket(final List<Integer> manualTicketNumbers) {
        List<LottoBall> lottoBalls = manualTicketNumbers.stream()
                .map(LottoBall::from)
                .collect(Collectors.toList());
        return new LottoTicket(new LottoBalls(lottoBalls));
    }
```

LottoMachine에서 LottoBalls에 List로 넘겨주기 위함입니다. (LottoBalls는 List로 들어온 인자값에 대해 유효성 검사를 진행함 )

> 혹시 이런경우 전체적으로 컬렉션 프레임워크를 맞춰야 하나요..?

choihz replied:
당연히 전체적으로 컬렉션 프레임워크를 맞춰야 한다! 그런 건 없죠. 그런데 동일 로직을 굳이 다르게 할 필요는 없지 않나 싶어 드린 코멘트였습니다. 수동 로또든 당첨 로또든 6개의 숫자를 입력받아 `LottoBalls` 객체가 되고, 이에 관련된 비즈니스 로직(중복 없어야 함, 6개여야 함...)은 동일하니까요.

Livenow14 replied:
여러가지를 사용해 보고 싶어서 이것저것 사용해봤어요..ㅎㅎ 

> 동일한 로직에선 동일한 프레임워크를 쓰는게 자연스럽다

이부분을 다음 미션을 진행하며 지키도록 하겠습니다.

좋은 지적 감사합니다 😃

choihz commented on src/main/java/domain/result/Result.java:
```diff
@@ -10,40 +10,34 @@
-        lottoRanks.forEach(this::putResult);
-        return results;
+    private Map<LottoRank, Integer> makeResult(final LottoTickets lottoTickets, final WinningLotto winningLotto) {
+        Map<LottoRank, Integer> copy = new HashMap<>();
```
`copy`라는 이름이 의미를 잘 드러내고 있나요?

Livenow14 replied:
의미를 잘 드러내지 않은거 같아요. 😭
result로 변경 후 의미를 더 잘 드러내게 하였습니다. 

0de24bf

choihz commented on src/main/java/domain/result/LottoRank.java:
```diff
@@ -20,12 +22,12 @@
-    public static LottoRank findRankByBonusAndMatches(final boolean isBonus, final int matches) {
-        if (!isBonus || matches != 5) {
+    public static LottoRank findRankByBonusAndMatches(final boolean hasBonus, final int matches) {
+        if (!hasBonus || matches != FIVE) {
```
`5`에 `FIVE`라는 이름을 붙이면 어떤 의미가 더해지나요? 🤔

Livenow14 replied:
의도를 잘 드러내지 않는거 같아. 
FIVE_MATCHES로 변경하였습니다 😊

5d0bcb2

choihz commented on src/test/java/domain/bettingMoney/BettingMoneyTest.java:
```diff
@@ -1,27 +1,74 @@
                 .doesNotThrowAnyException();
     }
 
     @DisplayName("보장된 숫자(로또 티켓 구입 금액보다 큰 양수의 숫자))가 아니면 에러가 발생한다.")
```
```suggestion
    @DisplayName("보장된 숫자(로또 티켓 구입 금액보다 큰 양수의 숫자)가 아니면 에러가 발생한다.")
```

Livenow14 replied:
확인했습니다 :)


23505ff

choihz commented on src/test/java/domain/LottoMachineTest.java:
```diff
@@ -0,0 +1,65 @@
+        Scanner scanner = new Scanner(inputStream);
+        InputView inputView = new InputView(scanner);
+        return inputView;
+    }
```
테스트를 위해 `InputView`를 직접 만들고 있군요.. 🤔

Livenow14 replied:
로직 수정으로 제거했습니다 :)

cd2ed3e

choihz reviewed (changes requested):
안녕하세요 검프!

2단계도 잘 구현해 주셨네요 👍 코멘트 보시고 궁금한 점, 의견 등은 1단계에서처럼 자유롭게 코멘트 또는 DM으로 알려주세요!

Livenow14 committed:
refactor: 결과 출력부분 메소드 분리

메소드를 분리하여 메소드의 책임을 나눴음

Livenow14 committed:
refactor: Stream 사용코드 변경

Stream을 사용해서 List화 했던 부분을 변경하였음.

Livenow14 committed:
refactor: BettingMoney 반환값 변경

getTicketCount가 바로 TicketCount 객체를 반환
생성 로직의 유효성 검사 변경, 메소드 내 유효성 검사 추가

Livenow14 committed:
refactor: LottoMachine::makeManualTickets 매개변수 변경

테스트 먼저 진행

makeManualTickets(InputView) -> makeManualTickets(List<Integer>)

makeManualTicket(InputView) -> makeManualTicket(List<Integer>)

Livenow14 committed:
refactor: 매개변수에 inputView 제거

controller에서 넘겨주던 inputView를 제거하고,
수동 로또 번호를 받는 책임을 Controller에서 수행하게 함.

LottoMachine::makeManualTickets 변경
LottoMachine::makeManualTicket 변경
LottoService::getLottoTickets 변경
GameManageApplication 내부 메소드 추가

Livenow14 committed:
refactor: Result::makeResult 지역변수 네이밍 변경

Livenow14 committed:
refactor: LottoRank 상수 네이밍 변경

FIVE -> FIVE_MATCHES

Livenow14 committed:
refactor: code convention

잘못된 이름 및 오타 수정

Livenow14 committed:
refactor: code convention

의미에 맞게 띄어쓰기를 함

Livenow14 committed:
refactor: LottoService 삭제

LottoMachine으로 책임이 충분할 것 같아 제거함.

Livenow14 commented:
안녕하세요 코니! 좋은 주말이에요!! `(주말에 리뷰하신다고 고생 많으십니다...ㅠㅠ)`
리팩터링 후 질문사항은 :eyes: 이모지로 표기했어요!!

이번 로또 미션에서 코니의 피드백을 통해 `책임을 생각하는 방법, 코드 작성의 이유, 라이브러리 사용의 이유` 등을 정말 깊게 생각해본 것 같아요!! 
항상 좋은 피드백을 남겨주셔서 감사합니다...! 😊😊 
이번 미션 이후에도 주신 피드백 잊지 않고 기억하겠습니다!! (더 성장해서 오겠습니다 😁)
이번에도 잘부탁드립니다..!

Livenow14 committed:
refactor: LottoRank 생성자 추가

생성자 추가를 통해 중복 제거

Livenow14 committed:
refactor: 상수 접근 지정자 변경

public으로 둘 이유가 없어, private으로 변경함.

choihz commented on src/main/java/view/LottoGameScreen.java:
```diff
@@ -14,37 +15,31 @@
-    public static final String REVENUE_RESULT_FORMATTER = "총 수익률은 %.2f입니다.";
+    private static final String LOTTO_TICKET_COUNT = "%d개를 구매했습니다.";
+    private static final String LOTTO_PREFIX = "[";
+    private static final String LOTTO_SUTFIX = "]";
```
오타가 있네요~

Livenow14 replied:
앗.. 너무 자연스러워서 몇번을 봤네요 
수정완료 했습니다 ㅎㅎ

choihz reviewed (approved):
안녕하세요 검프!

1단계에서부터 이어진 고된(?) 피드백의 루프를 끝낼 시간이 온 것 같습니다. 정말 정말 고생 많으셨어요! 로또 미션은 여기에서 마무리하도록 할게요. 다음에 또 만나요~ 👋

Livenow14 commented:
이번 로또 미션을 하며 정말 많은것을 생각할 수 있었습니다. 
제가 가지고 있던 습관들을 전부 다시 한번 돌이켜 봤어요. 
그중 가장 크게 배운것은 
>1.  현재  할수 잇는 일을 고민하라 (먼 미래의 확장성을 고려하지말자)
> 2. 코드 작성의 이유를 깊게 고민하라 (그냥 쓴다? nono~~ )
> 3. 통일성(대칭성)을 지키자

배웠던 것이 정말 많아서 그런지 전혀 고되지 않았어요! 
코니의 피드백을 통해 생각하는 방법을 배운 것 같아요. 

많은 질문, 많은 답변 하시느라 정말 고생많으셨습니다 다음에 뵙겠습니다 😊
